File: 351004/Beliaev/restapp/src/main/java/com/example/restapp/service/NoteService.java
package com.example.restapp.service;

import com.example.restapp.dto.request.NoteRequestTo;
import com.example.restapp.dto.response.NoteResponseTo;
import com.example.restapp.exception.EntityNotFoundException;
import com.example.restapp.mapper.NoteMapper;
import com.example.restapp.model.Article;
import com.example.restapp.model.Note;
import com.example.restapp.repository.ArticleRepository;
import com.example.restapp.repository.NoteRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class NoteService {
    private final NoteRepository noteRepository;
    private final ArticleRepository articleRepository;
    private final NoteMapper mapper;

    @Transactional
    public NoteResponseTo create(NoteRequestTo request) {
        Note note = mapper.toEntity(request);

        Article article = articleRepository.findById(request.getArticleId())
                .orElseThrow(() -> new EntityNotFoundException("Article not found with id: " + request.getArticleId()));
        note.setArticle(article);

        Note saved = noteRepository.save(note);
        return mapper.toResponse(saved);
    }

    public List<NoteResponseTo> getAll() {
        return noteRepository.findAll().stream()
                .map(mapper::toResponse)
                .collect(Collectors.toList());
    }

    public NoteResponseTo getById(Long id) {
        return noteRepository.findById(id)
                .map(mapper::toResponse)
                .orElseThrow(() -> new EntityNotFoundException("Note not found with id: " + id));
    }

    @Transactional
    public NoteResponseTo update(Long id, NoteRequestTo request) {
        Note note = noteRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Note not found with id: " + id));

        mapper.updateEntityFromDto(request, note);

        if (!note.getArticle().getId().equals(request.getArticleId())) {
            Article article = articleRepository.findById(request.getArticleId())
                    .orElseThrow(() -> new EntityNotFoundException("Article not found with id: " + request.getArticleId()));
            note.setArticle(article);
        }

        Note saved = noteRepository.save(note);
        return mapper.toResponse(saved);
    }

    @Transactional
    public void delete(Long id) {
        if (!noteRepository.existsById(id)) {
            throw new EntityNotFoundException("Note not found with id: " + id);
        }
        noteRepository.deleteById(id);
    }
}

File: 351002/Abrashin/discussion/src/main/kotlin/com/example/discussion/service/NoteService.kt
package com.example.discussion.service

import com.example.discussion.dto.request.NoteRequestTo
import com.example.discussion.dto.response.NoteResponseTo
import com.example.discussion.exception.BadRequestException
import com.example.discussion.exception.NotFoundException
import com.example.discussion.model.NoteEntity
import com.example.discussion.model.NoteKey
import com.example.discussion.repository.NoteRepository
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import kotlin.math.abs

@Service
class NoteService(
    private val noteRepository: NoteRepository
) {
    fun create(request: NoteRequestTo): NoteResponseTo {
        val tweetId = request.tweetId ?: throw BadRequestException("tweetId is required")
        val content = request.content ?: throw BadRequestException("content is required")
        val id = generateId()
        val saved = noteRepository.save(
            NoteEntity(
                key = NoteKey(tweetId = tweetId, id = id),
                country = request.country,
                content = content
            )
        )
        return toResponse(saved)
    }

    fun getById(id: Long): NoteResponseTo =
        toResponse(
            noteRepository.findByKeyId(id).firstOrNull()
                ?: throw NotFoundException("Note with id $id not found")
        )

    fun getAll(page: Int, size: Int, sort: Array<String>): List<NoteResponseTo> {
        val all = noteRepository.findAll().toList().map(::toResponse)
        val sorted = sortResults(all, sort)
        val from = (page * size).coerceAtMost(sorted.size)
        val to = (from + size).coerceAtMost(sorted.size)
        return sorted.subList(from, to)
    }

    fun put(id: Long, request: NoteRequestTo): NoteResponseTo {
        val existing = noteRepository.findByKeyId(id).firstOrNull()
            ?: throw NotFoundException("Note with id $id not found")
        val tweetId = request.tweetId ?: throw BadRequestException("tweetId is required")
        val content = request.content ?: throw BadRequestException("content is required")
        if (existing.key.tweetId != tweetId) {
            noteRepository.delete(existing)
        }
        val saved = noteRepository.save(
            NoteEntity(
                key = NoteKey(tweetId = tweetId, id = id),
                country = request.country,
                content = content
            )
        )
        return toResponse(saved)
    }

    fun patch(id: Long, request: NoteRequestTo): NoteResponseTo {
        val existing = noteRepository.findByKeyId(id).firstOrNull()
            ?: throw NotFoundException("Note with id $id not found")
        val newTweetId = request.tweetId ?: existing.key.tweetId
        val patched = existing.copy(
            key = NoteKey(tweetId = newTweetId, id = id),
            country = request.country ?: existing.country,
            content = request.content ?: existing.content
        )
        if (existing.key.tweetId != newTweetId) {
            noteRepository.delete(existing)
        }
        return toResponse(noteRepository.save(patched))
    }

    fun delete(id: Long) {
        val existing = noteRepository.findByKeyId(id).firstOrNull()
            ?: throw NotFoundException("Note with id $id not found")
        noteRepository.delete(existing)
    }

    fun getByTweetId(tweetId: Long): List<NoteResponseTo> =
        noteRepository.findByKeyTweetId(tweetId).map(::toResponse)

    private fun toResponse(entity: NoteEntity): NoteResponseTo = NoteResponseTo(
        id = entity.key.id,
        tweetId = entity.key.tweetId,
        country = entity.country,
        content = entity.content
    )

    private fun sortResults(items: List<NoteResponseTo>, sort: Array<String>): List<NoteResponseTo> {
        if (sort.isEmpty()) return items
        val field = sort.first()
        val direction = if (sort.size >= 2) Sort.Direction.fromString(sort[1]) else Sort.Direction.ASC
        val comparator = when (field) {
            "id" -> compareBy<NoteResponseTo> { it.id ?: 0L }
            "tweetId" -> compareBy<NoteResponseTo> { it.tweetId ?: 0L }
            else -> compareBy<NoteResponseTo> { it.id ?: 0L }
        }
        return if (direction.isAscending) items.sortedWith(comparator) else items.sortedWith(comparator.reversed())
    }

    private fun generateId(): Long = abs(java.util.UUID.randomUUID().mostSignificantBits)
}
